Initial support for optional navigation translation into LEFT OUTER JOIN.

maumar opened the pull request:
Currently we translate all navigations into INNER JOINs, which may produce invalid results for some queries.

Fix is to translate optional navigations into LEFT OUTER JOIN. In order to do that we convert optional navigation into SelectMany-GroupJoin-DefaultIfEmpty expression, that translates to LOJ.
We also need to add null checks in various places because now the intermittent results can be null.

e.g.
from c in ctx.Customers
where c.Detail.Name == "Foo"
select c

will get translated into:

from c in ctx.Customers
join d in ctx.Detail on c.Id equals d.CustomerId into grouping
from d in grouping.DefaultIfEmpty()
where (d != null ? d.Name : null) == "Foo"
select c;

Also fixed a number of bugs uncovered by this change, due to generating significantly more complex queries in some cases.

Known issues:
- DefaultIfEmpty() cannot be translated into SQL so everything happening after it is computed on the client (filters, projections, nested navigations) (#4588),
- Optional navigations don't work with Include (#4589),
- Optional navigations don't work with some queries involving SelectMany operator (#4539),
- Optional navigations don't work for soem complex queries involving subqueries and/or navigation inside inner key selector of a Join statement (#4547).

maumar commented:
@anpete  @divega

maumar commented:
new version up

maumar commented:
New version up

anpete commented on src/Microsoft.EntityFrameworkCore.Relational/Query/ExpressionVisitors/Internal/IsNullExpressionBuildingVisitor.cs:
```diff
@@ -9,13 +9,16 @@ namespace Microsoft.EntityFrameworkCore.Query.ExpressionVisitors.Internal
         protected override Expression VisitConstant(ConstantExpression node)
         {
-            if (node.Value == null)
+            if (node.Value == null && !_nullConstandAdded)
```
Typo

anpete commented:
@maumar Might be good if you link the known issues listed above to corresponding bugs.

maumar commented:
@anpete did that for some issues but I think not for all, will open issues for missing ones

maumar commented:
@anpete new version up, fixed the typo as well as some sql validation (which was broken on core clr), and added the bug numbers of the known issues in the description

anpete commented on src/Microsoft.EntityFrameworkCore/Query/ExpressionVisitors/Internal/NavigationRewritingExpressionVisitor.cs:
```diff
@@ -214,22 +287,40 @@ var rightNavigationJoin
             }
 
+            // TODO: should we special case other methods? e.g. Equals(object, object)?
+            if (node.NodeType != ExpressionType.ArrayIndex 
```
The hard-coded string.Concat methodinfo seems concerning. Can you elaborate?

maumar replied:
this is to unify the types of binary operation after nav rewrite. During the rewrite types get changed from non-nullable to nullable, e.g.

from c in ctx.Customers
where c.Detail.Id == 5
select c

this compensation needs to happen because c.Detail.Id will be changed to int?.

We take object into account also for cases where property is compared to null constant (which has it's type set to object). 

string.Concat is special cased for cases like:

from c in ctx.Customers
select c.Detail.Name + 1

This will use BinaryOperator with a method (string.Concat), and for this case we don't want to change the type of that second argument because it won't match the method signature.

However I think we can remove the special case for object and this method altogether, currently. This will cause null constant be evaluated on the client, but since everything is evaluated on the client anyway it doesn't make much difference. Once SelectMany-GroupJoin-DefaultIfEmpty is being properly translated to sql, we can re-evaluate this scenatio (we already have tests in place)

anpete commented:
:shipit:
